off_chain_data/application-go/parser/endorserTransaction.go

```
@@ -0,0 +1,115 @@
+package parser
+
+import (
+	"sync"
+
+	"github.com/hyperledger/fabric-protos-go-apiv2/ledger/rwset"
+	"github.com/hyperledger/fabric-protos-go-apiv2/peer"
+	"google.golang.org/protobuf/proto"
+)
+
+type endorserTransaction struct {
+	transaction   *peer.Transaction
+	readWriteSets func() ([]*readWriteSet, error)
+}
+
+func parseEndorserTransaction(transaction *peer.Transaction) *endorserTransaction {
+	result := &endorserTransaction{transaction, nil}
+	result.readWriteSets = sync.OnceValues(result.unmarshalReadWriteSets)
+	return result
+}
+
+func (p *endorserTransaction) unmarshalReadWriteSets() ([]*readWriteSet, error) {
+	chaincodeActionPayloads, err := p.unmarshalChaincodeActionPayloads()
+	if err != nil {
+		return nil, err
+	}
+
+	chaincodeEndorsedActions, err := p.extractChaincodeEndorsedActionsFrom(chaincodeActionPayloads)
+	if err != nil {
+		return nil, err
+	}
+
+	proposalResponsePayloads, err := p.unmarshalProposalResponsePayloadsFrom(chaincodeEndorsedActions)
+	if err != nil {
+		return nil, err
+	}
+
+	chaincodeActions, err := p.unmarshalChaincodeActionsFrom(proposalResponsePayloads)
+	if err != nil {
+		return nil, err
+	}
+
+	txReadWriteSets, err := p.unmarshalTxReadWriteSetsFrom(chaincodeActions)
+	if err != nil {
+		return nil, err
+	}
+
+	return p.parseReadWriteSets(txReadWriteSets), nil
+}
+
+func (p *endorserTransaction) unmarshalChaincodeActionPayloads() ([]*peer.ChaincodeActionPayload, error) {
+	result := []*peer.ChaincodeActionPayload{}
+	for _, transactionAction := range p.transaction.GetActions() {
+		chaincodeActionPayload := &peer.ChaincodeActionPayload{}
+		if err := proto.Unmarshal(transactionAction.GetPayload(), chaincodeActionPayload); err != nil {
+			return nil, err
+		}
+
+		result = append(result, chaincodeActionPayload)
+	}
+	return result, nil
+}
+
+func (*endorserTransaction) extractChaincodeEndorsedActionsFrom(chaincodeActionPayloads []*peer.ChaincodeActionPayload) ([]*peer.ChaincodeEndorsedAction, error) {
```

An error is never returned so you don't need the error return value. The results are just passed to `unmarshalProposalResponsePayloadsFrom` so you might find it easier to include this logic in that method. `chaincodeActionPayload.GetAction().GetProposalResponsePayload()`

---

This is great, done.